### PR TITLE
feat(RELEASE-2287): exclude non-binary files from signing in push-artifacts-to-cdn

### PR DESCRIPTION
**Pull request by swickersh**
Archives provided by product teams may contain non-binary files such as README.md or LICENSE alongside the actual binaries. Previously, the macOS and Windows signing steps would attempt to sign all files in the archive, causing failures on non-binary content.

This change updates the signing logic to skip files ending in .md or .txt:
- macOS: filter non-binary files from the find command before codesign
- Windows: use findstr to detect .md/.txt extensions and skip both signtool sign and signtool verify for those files

Non-binary files are preserved in the archive and flow through the rest of the pipeline (checksums, re-archiving, oras push, CGW publish) unchanged.

Includes a new unit test (test-push-artifacts-to-cdn-non-binary-files-in-archive) with supporting mocks to validate the behavior.

Assisted-by: Cursor AI

## Checklist before requesting a review
- [x] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [x] My commit message includes `Signed-off-by: My name <email>`
- [x] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)
- [ ] I have run the README.md generator script in `.github/scripts/readme_generator.sh` and verified the results using `.github/scripts/check_readme.sh`

### CODE REVIEW

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 1]
```diff
@@ -725,7 +725,7 @@ spec:
           security unlock-keychain -p $KEYCHAIN_PASSWORD login.keychain
           set -x
           echo "Signing files in the \$CONTENT_DIR_MAC directory..."
-          find "\$CONTENT_DIR_MAC" -type f | while read file; do
+          find "\$CONTENT_DIR_MAC" -type f ! -name '*.md' ! -name '*.txt' | while read file; do
```
Comment (deekej) [inline tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: You might want to include the `LICENSE` file here as well - many projects use it without a suffix.

Also, it's possible for some products to use capital `*.MD` or `*.TXT` suffixes.

Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: Done. Thanks for pointing out

**Comment by swickersh** [comment]
> ## PR Reviewer Guide 🔍
> Here are some key observations to aid the review process:

> [**Edge Case**](https://github.com/konflux-ci/release-service-catalog/pull/2021/files#diff-5b8756a69e7d7bf343436372e3a297add65fb59fe212ee781a3b91993fd62f14R889-R917)
> Windows skip logic claims to skip `LICENSE*` but the condition only skips when the base name is exactly `LICENSE`. Files like `LICENSE.txt`, `LICENSE-MIT`, or `LICENSE.md` would still be signed/verified. If the intention is truly `LICENSE*`, adjust the matching (e.g., use `if /I "%%~nxf"=="LICENSE"` for exact, or pattern-based matching via `findstr`/wildcards) to align behavior with the comment.
> 

Qodo is incorrect.
Anything .txt or .md is skipped (this includes LICENSE.md)

LICENSE-MIT would cause a failure. But we have to draw the line somewhere. I don't want to allow signing to be skipped for any arbitrary file name with no extension.  I will make sure the documentation is clear on the allowed supplementary files and the strict naming/extensions.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 2]
```diff
@@ -725,7 +725,7 @@ spec:
           security unlock-keychain -p $KEYCHAIN_PASSWORD login.keychain
           set -x
           echo "Signing files in the \$CONTENT_DIR_MAC directory..."
-          find "\$CONTENT_DIR_MAC" -type f | while read file; do
+          find "\$CONTENT_DIR_MAC" -type f ! -iname '*.md' ! -iname '*.txt' ! -iname 'license' | while read file; do
```
[the remark this answers is not in the document]
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: filenames with spaces and special chars are not allowed.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 3]
```diff
@@ -725,7 +725,7 @@ spec:
...
               echo "Signing: \$file"
               if ! xcrun codesign --sign "Developer ID Application: $SIGNING_IDENTITY" \
                   --options runtime --timestamp --force "\$file"; then
```
[the remark this answers is not in the document]
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml:779]: Done

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 4]
```diff
@@ -725,7 +725,12 @@ spec:
           security unlock-keychain -p $KEYCHAIN_PASSWORD login.keychain
           set -x
           echo "Signing files in the \$CONTENT_DIR_MAC directory..."
-          find "\$CONTENT_DIR_MAC" -type f | while read file; do
+          # NOTE: We skip signing non-binary documentation/license files
+          # (e.g., *.md, *.txt, LICENSE) because codesigning expects binaries
+          # and fails on text files.
+          # This only affects signing/verification; these files still flow
+          # through packaging/checksums/publishing.
+          find "\$CONTENT_DIR_MAC" -type f ! -iname '*.md' ! -iname '*.txt' ! -iname 'license' | while read file; do
```
Comment (elenagerman) [inline tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: I see QoDo already proposed it, but you rejected it with the "filenames with spaces and special chars are not allowed."
Just to clarify: not allowed by what?
schema/dataKeys.json defines files[].source and staged.files[].source as plain "type": "string" with no pattern constraint. There is nothing that rejects spaces or special characters.
What the README says — nothing about filename restrictions.
What the task code does — no validation step rejects unusual filenames in the archive contents. The archives come from product teams; the pipeline just unpacks whatever is inside them with tar -xf.
in practice, release binary filenames from Red Hat product teams follow a conventional pattern and spaces are essentially never used. The risk in production is very low.
And I agree that it is not a "bug", but more like "robustness/best practice". The -r flag is still worth adding; maybe you can reconsider it.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 5]
```diff
@@ -886,21 +891,39 @@ spec:
...
           for /r unsigned\windows %%f in (*) do (
-            signtool sign /v /n "Red Hat" /fd SHA256 /tr http://timestamp.digicert.com /td SHA256 "%%f"
-            if errorlevel 1 (
-              echo Signing of %%f failed
-              exit /B %ERRORLEVEL%
+            if /I "%%~xf"==".md" (
+              echo Skipping non-binary file: %%f
+            ) else if /I "%%~xf"==".txt" (
+              echo Skipping non-binary file: %%f
+            ) else if /I "%%~nf"=="LICENSE" (
```
Comment (elenagerman) [inline tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: there is inconsistency with the comment: LICENSE* vs LICENSE.
this matches only a file whose base name is exactly LICENSE, but not wildcard.
for example:  LICENSE-MIT has base name LICENSE-MIT, not LICENSE, so it would not be skipped.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml
Comment (elenagerman) [inline tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: it became to be a dead code and  misleading
If any file's signtool verify fails, the inner loop already calls exit /B %ERRORLEVEL% immediately (line 907 and 925), so execution never reaches this block in this case. If all verifications pass, errorlevel is 0 when the loop ends, so this block is never entered either.

Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: yep - that's dead now. Removed. Thanks

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml (hunk 4 above)
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: >schema/dataKeys.json defines files[].source and staged.files[].source as plain "type": "string" with no pattern constraint. There is nothing that rejects spaces or special characters.

These supplementary files aren't explicitly declared in the RPA. So datakeys.json won't be able to gate them.  Part of the Jira AC includes me updating the user docs which will instruct them on the file naming restrictions agreed with relengs. If users don't abide by the naming conventions, the code will currently fail fast as their supplementary files (license and readmes) would attempt signing and fail. 

But I see an area for improvement here. Thanks for calling out - I'll create an allowed_supplementary files list and gate them in push-unsigned-using-oras step with a sanitizer. That will ensure only files explicitly listed in RPAs or files following our naming conventions are iterated on by the rest of the pipeline steps. Also adding the -r flag even though it's not really necessary anymore with the sanitizer, but it's zero-cost defensive.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml (hunk 5 above)
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: fixed comment. It's also addressed with the sanitizer.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 6]
```diff
@@ -455,6 +455,35 @@ spec:
...
+              for a in "${ALLOWED_SUPPLEMENTARY[@]}"; do
+                [[ "$name" == "$a" ]] && { allowed=true; break; }
+              done
+              if [[ "$allowed" == false ]]; then
+                echo "  Removing non-allowlisted supplementary file: $name"
+                rm -f "$file"
+              fi
+            fi
+          done < <(find "$dir" -type f -print0)
+        }
```
[the remark this answers is not in the document]
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: codesign on Mac fails on non-binary files.
signtool on Windows fails on non-PE files.
Both fail if the file path doesn't resolve (broken symlink).
So the sanitizer handles the common/expected cases cleanly (wrong filenames, non-allowlisted supplementary files), and anything bizarre that slips through would just cause a signing failure, which stops the pipeline and surfaces the problem. Nothing gets silently published with bad content.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml (hunk 4 above)
Comment (elenagerman) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: thank you!

**Review by elenagerman — DISMISSED** [review]

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 7]
```diff
@@ -455,6 +455,35 @@ spec:
         # due to set -e, this catches all EXIT and ERR calls and the task should never fail with nonzero exit code
         trap 'exitfunc $? $LINENO "$BASH_COMMAND"' EXIT
 
+        ALLOWED_SUPPLEMENTARY=(
+          "README.md" "README.MD" "readme.md" "readme.MD"
```
Comment (davidmogar) [inline tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: So, I'm getting correctly that _only_ this files need to be skipped? What about `LICENSE.MD`, `Readme.md` or even `CHANGELOG.MD`? Would those be fine?

If you need to skip anything not being a binary, can `file` output be used?

Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: You understand it correctly and only the 8 filenames in allowed_supplementary are eligible to skip signing.  The rationale behind this is to err on the side of caution in concern of files intentionally or non-intentionally skipping the signing workflow.
The documentation update from the AC of the Jira will be clear on the file naming restrictions of the supplementary files.  If teams request changelogs or additional supplementary files, we can always address those later.

It could also be considered that the allowed_supplementary is propagated from an RPA field or controlled by the constraints file in the KRD repo somehow.  But if that RFE comes at some point, it would be simple to adjust this to read from the data.json instead of the list defined here.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 8]
```diff
@@ -455,6 +455,35 @@ spec:
         # due to set -e, this catches all EXIT and ERR calls and the task should never fail with nonzero exit code
         trap 'exitfunc $? $LINENO "$BASH_COMMAND"' EXIT
 
+        ALLOWED_SUPPLEMENTARY=(
```
Comment (johnbieren) [inline tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: I'm a bit confused by this. I get not wanting to sign text files. But on line 481 we are only removing the files if they do not have one of these names? Shouldn't we be removing these too so they aren't signed? Why are these special and not like CLAUDE.md or CODE_OF_CONDUCT.md?

Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: I agree this code isn't reading as intuitively as I hoped.
Refactoring to a similar idea, but simpler approach.  We'll simply filter out the allowed supplementary files to ensure the signing hosts don't attempt to sign them (because signing will fail) and we'll restore the files after signing.
Due to the RPA only having the archive names and not the expected filenames within the archive, we don't have a straightforward way to filter out supplementary files that aren't in our allowed list.  
IMO, the cleanest way to handle this would just be fail early when a product team intentionally or unintentionally includes a non-allowed file in their release archive. That earliest point would be during the signing step and the error message will say which file failed to sign.
As of now, I've only heard of teams wanting to include a LICENSE and README file. If the list grows, then I'd suggest we consider adding the supplementary_files list to the RPA and maybe prodsec can govern that with constraints or something. It would be an easy refactor here later.

Lastly - I tweaked the code a bit so I'll run some more test pipelines now. Moving back to draft for a couple hours.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 9]
```diff
@@ -725,7 +776,7 @@ spec:
           security unlock-keychain -p $KEYCHAIN_PASSWORD login.keychain
           set -x
           echo "Signing files in the \$CONTENT_DIR_MAC directory..."
-          find "\$CONTENT_DIR_MAC" -type f | while read file; do
+          find "\$CONTENT_DIR_MAC" -type f | while IFS= read -r file; do
               echo "Signing: \$file"
               if ! xcrun codesign --sign "Developer ID Application: $SIGNING_IDENTITY" \
                   --options runtime --timestamp --force "\$file"; then
```
[the remark this answers is not in the document]
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml:779]: pre-existing and low risk

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 10]
```diff
@@ -1096,7 +1142,22 @@ spec:
...
+          for os_name in macos windows; do
+            [ -d "$SUPP_HOLD/$os_name" ] || continue
+            while IFS= read -r -d '' file; do
+              rel="${file#"$SUPP_HOLD/$os_name"/}"
+              dest="$SIGNED_DIR/$os_name/$rel"
+              mkdir -p "$(dirname "$dest")"
+              mv "$file" "$dest"
+              echo "  Restored supplementary file: $os_name/$rel"
+            done < <(find "$SUPP_HOLD/$os_name" -type f -print0)
+          done
```
[the remark this answers is not in the document]
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml:1153]: For the collision to occur you'd need one of:

- A binary is literally named README.md — someone packed an executable with that name into the archive. Possible in theory but would already be a broken release package.
- The oras pull for signed artifacts somehow also pulls a supplementary file — that would require the signing host to have added one, which it has no reason to do.
- $SIGNED_DIR already contained files from a previous loop iteration — but COMPONENT_DIR is unique per component so there's no cross-contamination.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 11]
```diff
@@ -455,6 +455,52 @@ spec:
...
+          while IFS= read -r -d '' file; do
+            if is_supplementary_file "$file"; then
+              local rel="${file#"$src_root"/}"
+              local dest="$hold_root/$rel"
+              mkdir -p "$(dirname "$dest")"
+              mv "$file" "$dest"
+              echo "  Held supplementary file: $rel"
+            fi
+          done < <(find "$src_root" -type f -print0)
+        }
```
[the remark this answers is not in the document]
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml:500]: duplicate - addressed in another comment

File: tasks/internal/push-artifacts-to-cdn-task/tests/mocks.sh [hunk 12]
```diff
@@ -78,6 +78,40 @@ EOF
...
+        create_mock_tgz_with_non_binary_files() {
+            local filename="$1"
+            local binary_name="${filename%.tar.gz}"
+
+            local temp_bin_dir=$(mktemp -d)
+            echo "Mock binary content for $binary_name" > "$temp_bin_dir/$binary_name"
+            chmod +x "$temp_bin_dir/$binary_name"
+            echo "README for $binary_name" > "$temp_bin_dir/README.md"
+            echo "License text for $binary_name" > "$temp_bin_dir/LICENSE"
+            echo "Changelog for $binary_name" > "$temp_bin_dir/CHANGELOG.txt"
```
[the remark this answers is not in the document]
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/tests/mocks.sh:90]: done.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 13]
```diff
@@ -455,6 +455,52 @@ spec:
         # due to set -e, this catches all EXIT and ERR calls and the task should never fail with nonzero exit code
         trap 'exitfunc $? $LINENO "$BASH_COMMAND"' EXIT
 
+        # Supplementary file classification (e.g. readme, license, changelog)
+        # Any file matching this is moved out before signing and restored afterward. Any other non-binary file will 
+        # cause the signing tool to fail with a clear error, which is the intended behavior. We woud ideally fail
```
Comment (johnbieren) [inline tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: s/woud/would/

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml (hunk 8 above)
Comment (johnbieren) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: Code looks a lot more intuitive to me now, thanks

**Comment by davidmogar** [comment]
This PR is in draft so I marked it as not ready for review. Please update the sheet once ready.

**Review by midnightercz — DISMISSED** [review]

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml (hunk 13 above)
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml]: done

**Comment by swickersh** [comment]
> ## PR Reviewer Guide 🔍
> Here are some key observations to aid the review process:
> 
> **🎫 Ticket compliance analysis 🔶**
> 
> **[RELEASE-2191](https://redhat.atlassian.net/browse/RELEASE-2191) - Partially compliant**
> 

RELEASE-2191 was later split into a second Jira RELEASE-2287. This complies with the AC. The base branch name is probably confusing the bot. 



[RELEASE-2191]: https://redhat.atlassian.net/browse/RELEASE-2191?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

File: tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh
[the remark this answers is not in the document]
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh:1]: We've touched on this already and decided we'll make it configurable later if teams request more supplementary files. It could potentially be passed from the RPA

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 14]
```diff
@@ -1659,7 +1720,11 @@ spec:
             push_component_to_pulp "$COMPONENT_NAME" 2> >(tee "$STDERR_FILE" >&2)
           elif [[ "$HAS_CGW" == "true" ]]; then
             # Only push to CDN when NOT pushing to pulp
-            push_component_to_cdn "$COMPONENT_NAME" > >(tee "$STDERR_FILE") 2>&1 || true
+            push_component_to_cdn "$COMPONENT_NAME" 2>&1 | tee "$STDERR_FILE"
+            cdn_exit=${PIPESTATUS[0]}
+            if [ "$cdn_exit" -ne 0 ]; then
+              exit "$cdn_exit"
+            fi
```
[the remark this answers is not in the document]
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml:1720]: The EXIT trap is not "silently ignoring" CDN failures -- it's the mechanism that records them. Disabling the trap before exiting would lose the error details and produce a worse failure mode.

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml [hunk 15]
```diff
@@ -574,6 +620,11 @@ spec:
             rm -f "$ARCHIVE_PATH"
           done
 
+          SUPP_HOLD="$COMPONENT_DIR/supplementary"
+          echo "Moving supplementary files out of signing directories..."
+          move_supplementary_out "$UNSIGNED_DIR/macos" "$SUPP_HOLD/macos"
```
Comment (johnbieren) [inline tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml:623]: I see the tests pass, but why isn't it making you `mkdir` these? just for my own curiosity

File: tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh
Comment (johnbieren) [inline tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh:1]: (Feel free to resolve this without action) This is definitely a good file to have, but I don't think it will be executed in CI
https://github.com/konflux-ci/release-service-catalog/blob/development/.github/scripts/test_tekton_tasks.sh#L181
Probably not worth changing the wiring to make it happen though, since we will be moving the scripts out of here and changing to python, where we can have plenty of unit tests, anyways

**Review by johnbieren — DISMISSED** [review]

File: tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml (hunk 15 above)
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/push-artifacts-to-cdn-task.yaml:623]: it does on line 497

File: tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh:1]: Oh good point. Yea I only ran it locally. And yes, at this point would rather just wait until the python conversion.

**Review by davidmogar — DISMISSED** [review]

**Review by ach912 — APPROVED** [review]

File: tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+#   move_supplementary_out()
+#
+# Run directly:  bash test-supplementary-file-classification.sh
+#
+set -euo pipefail
+
+PASS=0
+FAIL=0
+
+# ── helpers under test (mirrored exactly from the task) ──────────────────
```
Comment (ralphbean) [inline tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh:16]: Note (defer): These helper functions are copy-pasted from the task YAML (as the comment acknowledges). If `is_supplementary_file()` or `move_supplementary_out()` are modified in the task, these tests will silently pass against the old logic. Consider adding a `# SYNC-CHECK:` marker referencing the source, or a CI check that diffs the two copies to detect drift.

File: tasks/internal/push-artifacts-to-cdn-task/tests/test-push-artifacts-to-cdn-non-binary-files-in-archive.yaml [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+              #!/usr/bin/env bash
+              set -ex
+
+              # Supplementary files (README.md, LICENSE, CHANGELOG.txt) are
+              # moved out before signing and restored afterward. They flow
+              # through to checksums, compression, and publishing.
+              if [[ "$(params.result)" != "Success" ]]; then
+                echo "Error: expected Success but got: $(params.result)"
+                echo "Archives with supplementary files should not break the pipeline"
+                exit 1
```
Comment (ralphbean) [inline tasks/internal/push-artifacts-to-cdn-task/tests/test-push-artifacts-to-cdn-non-binary-files-in-archive.yaml:99]: Note (defer): This test only checks `result == "Success"`, which proves the pipeline does not crash but not that supplementary files were actually moved out and restored. A pipeline that silently skips the supplementary logic would also pass. Consider grepping the task logs for the "Held supplementary file:" and "Restored supplementary file:" echo messages to strengthen validation. Same applies to the sanitize-filenames test.

File: tasks/internal/push-artifacts-to-cdn-task/tests/mocks.sh [hunk 18]
```diff
@@ -78,6 +78,40 @@ EOF
...
+            local temp_bin_dir=$(mktemp -d)
+            echo "Mock binary content for $binary_name" > "$temp_bin_dir/$binary_name"
+            chmod +x "$temp_bin_dir/$binary_name"
+            echo "README" > "$temp_bin_dir/README.md"
+            echo "License" > "$temp_bin_dir/LICENSE"
+            echo "changelog" > "$temp_bin_dir/CHANGELOG.md"
+            # Invalid: space in filename — will be removed by sanitizer
+            echo "bad" > "$temp_bin_dir/bad file.txt"
+
+            tar -czf "temp_releases/releases/$filename" -C "$temp_bin_dir" \
```
Comment (ralphbean) [inline tasks/internal/push-artifacts-to-cdn-task/tests/mocks.sh:110]: Note (defer): This comment says "will be removed by sanitizer", but the pipeline test description says "Unsafe-named files are not removed; they are passed through to the signing host unchanged." There is no sanitizer in the task itself. These statements contradict each other — worth aligning to avoid misleading future contributors.

**Review by ralphbean — DISMISSED** [review]
Good work on solving the supplementary file signing problem. The approach of moving non-binary files out before signing and restoring them afterward is sound, and the classification logic is well-structured with appropriate case-insensitive matching.

A few notes for future consideration (none blocking):
- The unit test copies helper functions from the task YAML — consider adding a drift-detection mechanism.
- The pipeline integration tests validate "Success" but don't verify the move/restore actually happened.
- A mock comment contradicts the test description regarding a "sanitizer" — worth aligning.

The CDN error handling improvement (`|| true` → `PIPESTATUS`), the macOS `read` robustness fix, and the Windows dead-code removal are all welcome improvements.

File: tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh (hunk 16 above)
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/tests/test-supplementary-file-classification.sh:16]: Added #sync-check comment

File: tasks/internal/push-artifacts-to-cdn-task/tests/test-push-artifacts-to-cdn-non-binary-files-in-archive.yaml (hunk 17 above)
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/tests/test-push-artifacts-to-cdn-non-binary-files-in-archive.yaml:99]: thanks, implemented this.

File: tasks/internal/push-artifacts-to-cdn-task/tests/mocks.sh (hunk 18 above)
Comment (swickersh) [reply tasks/internal/push-artifacts-to-cdn-task/tests/mocks.sh:110]: thanks, "sanitizer" was a remnant from a previous approach I took.  Reworded throughout to better represent what this does.

**Review by ralphbean — APPROVED** [review]
Clean approach to the supplementary file handling. The hold-out/restore pattern is well-designed — it keeps the classification logic centralized and avoids scattering extension checks across the signing scripts. The unit tests are thorough with good edge case coverage, and the integration tests follow existing patterns.

The CDN push error handling change (removing `|| true`) is a welcome improvement — silent failures are worse than loud ones.

LGTM.
